virtcontainers/kata_agent.go

```diff
@@ -1294,7 +1294,7 @@ func (k *kataAgent) handleLocalStorage(mounts []specs.Mount, sandboxID string) [
 			// We rely on the fact that the first container in the VM has the same ID as the sandbox ID.
 			// In Kubernetes, this is usually the pause container and we depend on it existing for
 			// local directories to work.
-			mounts[idx].Source = filepath.Join(kataGuestSharedDir, sandboxID, KataLocalDevType, filepath.Base(mnt.Source))
+			mounts[idx].Source = filepath.Join(kataGuestSharedDir, sandboxID, rootfsSuffix, KataLocalDevType, filepath.Base(mnt.Source))
```

Comment:
So the design is to translate empty dir media type "local" to mean to use container rootfs storage, right?

For dm, we have to create this inside container rootfs because we don't have out-of-rootfs place to create an empty dir as mount source. For overlayfs kind of rootfs, we can just mount the k8s empty dir source to kataGuestShareDir. Any reason that is not chosen?

Reply:
@bergwolf The idea is to create the empty-dir on the sandbox rootfs and make this behaviour identical in case of overlay and devicemapper. While we were doing this for devicemapper, due to the way the paths are constructed, we were not creating the empty-dir on the sandbox rootfs in case of overlay.
We decided to go this way to benefit from the performance of devicemapper and to keep the behaviour identical for all storage drivers. I dont think we should look at what storage driver is being used for the container rootfs to determine the behaviour for other volumes.
The plan is to eventually get rid of this special handling and simply pass the host-created directories to the guest, once we move virtio-fs as a stable feature.

Reply:
@amshinde How is virtio-fs going to help the situation? I think it is just a replacement for better 9pfs. The hack here was introduced for devicemapper because containerd snapshotter does not create `_rootfs` dir and thus we have to create a real empty dir on the container rootfs to be able to use the devicemapper storage.